File: Code/src/domain/protocols.py
from pathlib import Path
from typing import TYPE_CHECKING, Protocol
from .entities import PredictionResult

if TYPE_CHECKING:
    from .treatment import TreatmentInfo

class ImageClassifier(Protocol):
    def classify(self, image_path: Path) -> PredictionResult:
        ...

class TreatmentRepository(Protocol):
    def get_by_label(self, label: str) -> "TreatmentInfo | None":
        ...

File: Code/lib/domain/Protocols.dart
import 'package:image_picker/image_picker.dart';
import 'Entities.dart';
import 'Treatment.dart';
import 'ZoneAnalysis.dart';
import 'ClimateData.dart';
import 'OnsetEstimate.dart';

abstract class ImageClassifier {
  Future<PredictionResult> classify(XFile image);
}

abstract class TreatmentRepository {
  TreatmentInfo? getByLabel(String label);
  TreatmentInfo? getByLabelAndSeverity(String label, String severityLevel);
}

abstract class ZoneAnalyzer {
  Future<ZoneAnalysis> analyze(XFile image, {double? lat, double? lon});
}

abstract class ClimateRepository {
  Future<ClimateData?> fetch(double lat, double lon);
}

abstract class OnsetEstimator {
  OnsetEstimate estimate({
    required String pathogenClass,
    required String severityLevel,
    ClimateData? climate,
  });
}
